### PR TITLE
implemented TPM normalization

### PR DESCRIPTION
**Pull request by moritzschaefer**
This is an early draft. What needs to be discussed?
- tpm works different from cpm in that it does not change the internal counts_variable. The way cpm is working confused me: The counts array is normalized and the original data is lost (overwritten). I think the name "cpm" rather suggests to return cpm normalized values. 
One way to make this clear is for example to name the function
normalize_by_cpm or so (note the imperative). 

In my opinion it feels wrong not to keep the original counts data (if a user performs several subsequent normalizations (first cpm, then tpm) it might be an issue so I suggest the following: 

We have three properties:
- counts
- normalized_counts
- normalization_methods

The methods 
- cpm
- tpm
- rpkm
do return the normalized counts (but not alter the normalized_counts property). An additional method
- normalize(method: str) (method is one 'cpm', 'tpm', 'rpkm')
(note the imperative again), which is used to alter the internal normalized_counts field. 

The normalized_counts field is used in the object for all other computations (as is counts at the moment). 

Furthermore there are no tests at the moment. I first wanted to start discussions.

### CODE REVIEW

File: edgePy/DGEList.py [hunk 1]
```diff
@@ -292,17 +292,40 @@ def rpkm(
...
-        # TODO: Implement here
+        # based on https://haroldpimentel.wordpress.com/2014/05/08/what-the-fpkm-a-review-rna-seq-expression-units/ and
+        # https://gist.github.com/slowkow/c6ab0348747f86e2748b
+        # how many counts per base
 
-        # self = self.cpm(log=log, prior_count=prior_count)
+        # compute effective length not allowing negative lengths
+        if mean_fragment_lengths:
+            effective_lengths = (gene_lengths[:, None] - mean_fragment_lengths[None, :]).clip(
+                min=1
```
Comment (moritzschaefer) [inline edgePy/DGEList.py]: A negative length would not make sense and clipping to 0 would result in division by zeros.

File: edgePy/DGEList.py [hunk 2]
```diff
@@ -292,17 +292,40 @@ def rpkm(
...
 
-        # TODO: Implement here
+        # based on https://haroldpimentel.wordpress.com/2014/05/08/what-the-fpkm-a-review-rna-seq-expression-units/ and
+        # https://gist.github.com/slowkow/c6ab0348747f86e2748b
+        # how many counts per base
 
-        # self = self.cpm(log=log, prior_count=prior_count)
+        # compute effective length not allowing negative lengths
+        if mean_fragment_lengths:
+            effective_lengths = (gene_lengths[:, None] - mean_fragment_lengths[None, :]).clip(
```
Comment (clintval) [inline edgePy/DGEList.py]: What do you think about this version for readability sake?

```python
effective_lengths = np.subtract(gene_lengths, mean_fragment_lengths).clip(min=1)
```

Or:

```python
effective_lengths = (gene_lengths - mean_fragment_lengths).clip(min=1)
```

Why did you need to slice?

File: edgePy/DGEList.py [hunk 3]
```diff
@@ -292,17 +292,40 @@ def rpkm(
...
+            log: store log outputs
+            prior_count:
+            mean_fragment_lengths: vector of mean fragment length for each sample in counts table (optional)
 
         """
-        raise NotImplementedError
 
-        # TODO: Implement here
+        # based on https://haroldpimentel.wordpress.com/2014/05/08/what-the-fpkm-a-review-rna-seq-expression-units/ and
+        # https://gist.github.com/slowkow/c6ab0348747f86e2748b
```
Comment (clintval) [inline edgePy/DGEList.py]: Use the formulation or derivation as published and cite that:

```
Wagner, et al. Measurement of mRNA abundance using RNA-seq data:
RPKM measure is inconsistent among samples. Theory Biosci. 24 July 2012.
doi:10.1007/s12064-012-0162-3
```

Downside of citing other algorithm implementations (from source or gists) is that you inherit their assumptions (and bugs) even if you don't intend to.

File: edgePy/DGEList.py [hunk 4]
```diff
@@ -292,17 +292,40 @@ def rpkm(
...
+        self,
+        gene_lengths: np.ndarray,
+        log: bool = False,
+        prior_count: float = PRIOR_COUNT,
+        mean_fragment_lengths: np.ndarray = None,
+    ) -> np.ndarray:
+        """Normalize the DGEList to transcripts per million
+
+        Args:
+            gene_lengths: vector of length for each gene in counts table
```
Comment (clintval) [inline edgePy/DGEList.py]: Python doesn't really have a `vector` type in NumPy-speak. Maybe:

```
gene_lengths: 1D array of gene lengths for each gene in the rows of `DGEList.counts`.
mean_fragment_lengths: 1D array of mean fragment lengths for each sample in the columns of  `DGEList.counts` (optional)
```

Up to you.

File: edgePy/DGEList.py [hunk 5]
```diff
@@ -292,17 +292,40 @@ def rpkm(
...
-    ) -> None:
-        """Return the DGEList normalized to reads per kilobase of transcript
-        length.
+        self,
+        gene_lengths: np.ndarray,
+        log: bool = False,
+        prior_count: float = PRIOR_COUNT,
+        mean_fragment_lengths: np.ndarray = None,
+    ) -> np.ndarray:
+        """Normalize the DGEList to transcripts per million
```
Comment (clintval) [inline edgePy/DGEList.py]: Not quite. The way you have the function work now is that you are returning the normalized counts matrix (2D array) and not the normalized `DGEList` object.

File: edgePy/DGEList.py [hunk 6]
```diff
@@ -292,17 +292,40 @@ def rpkm(
...
+            effective_lengths = (gene_lengths[:, None] - mean_fragment_lengths[None, :]).clip(
+                min=1
+            )
+        else:
+            effective_lengths = gene_lengths[:, None]
+
+        base_counts = self.counts / effective_lengths
+
+        self.counts = 10 ** 6 * base_counts / np.sum(base_counts, axis=0)[None, :]
+        if log:
```
Comment (clintval) [inline edgePy/DGEList.py]: We don't assert `log` is a boolean so to be safe I always do:


```python
if log is True:
    ...
```

Otherwise other truthy values can slip in if the user is confused about what `log` is.

For example, this would be a valid value for `log` in this function:

```python
log = 1e10
```

File: edgePy/DGEList.py [hunk 7]
```diff
@@ -292,17 +292,40 @@ def rpkm(
...
+        gene_lengths: np.ndarray,
+        log: bool = False,
+        prior_count: float = PRIOR_COUNT,
+        mean_fragment_lengths: np.ndarray = None,
+    ) -> np.ndarray:
+        """Normalize the DGEList to transcripts per million
+
+        Args:
+            gene_lengths: vector of length for each gene in counts table
+            log: store log outputs
```
Comment (clintval) [inline edgePy/DGEList.py]: I strongly prefer qualifying this variable name with verb or preposition:

So how about `to_log_transform` over `log` which indicates the option is a boolean choice and that the action is an _in situ_ transformation.

https://github.com/r-bioinformatics/edgePy/issues/23#issuecomment-404251095

Comment (clintval) [reply edgePy/DGEList.py]: Could you change elsewhere in the codebase so we are consistent?

File: edgePy/DGEList.py [hunk 8]
```diff
@@ -292,17 +292,40 @@ def rpkm(
...
+            prior_count:
```
Comment (clintval) [inline edgePy/DGEList.py]: Need a definition.

File: edgePy/DGEList.py [hunk 9]
```diff
@@ -292,17 +292,40 @@ def rpkm(
...
-        raise NotImplementedError
 
-        # TODO: Implement here
+        # based on https://haroldpimentel.wordpress.com/2014/05/08/what-the-fpkm-a-review-rna-seq-expression-units/ and
+        # https://gist.github.com/slowkow/c6ab0348747f86e2748b
+        # how many counts per base
 
-        # self = self.cpm(log=log, prior_count=prior_count)
+        # compute effective length not allowing negative lengths
+        if mean_fragment_lengths:
```
Comment (clintval) [inline edgePy/DGEList.py]: Since we do not validate this arg. I suggest being more explicit:

```python
if mean_fragment_lengths is not None:
    ...
```

This saves the user from confusing stack traces when they pass in a value like:

```python
mean_fragment_lengths = 2
```

Instead of:

```python
mean_fragment_lengths = np.array([2])
```

File: edgePy/DGEList.py [hunk 10]
```diff
@@ -292,17 +292,40 @@ def rpkm(
         # self = self.cpm(log=log, prior_count=prior_count)
 
     def tpm(
-        self, transcripts: Mapping, log: bool = False, prior_count: float = PRIOR_COUNT
-    ) -> None:
-        """Return the DGEList normalized to reads per kilobase of transcript
-        length.
+        self,
+        gene_lengths: np.ndarray,
```
Comment (clintval) [inline edgePy/DGEList.py]: We need an assertion (and raise exception) if this is not a np.array and the shape of this array is not 1D.

Do something like:

```python
if not isinstance(gene_lengths, np.ndarray):
    raise TypeError('...')
if gene_lengths.shape() != (1, 0):
    raise ValueError('...')
```

Do the same for `mean_fragment_lengths`.

File: edgePy/DGEList.py [hunk 11]
```diff
@@ -292,17 +292,40 @@ def rpkm(
...
+        # how many counts per base
 
-        # self = self.cpm(log=log, prior_count=prior_count)
+        # compute effective length not allowing negative lengths
+        if mean_fragment_lengths:
+            effective_lengths = (gene_lengths[:, None] - mean_fragment_lengths[None, :]).clip(
+                min=1
+            )
+        else:
+            effective_lengths = gene_lengths[:, None]
```
Comment (clintval) [inline edgePy/DGEList.py]: Why slice?

File: edgePy/DGEList.py [hunk 12]
```diff
@@ -292,17 +292,40 @@ def rpkm(
...
+
+        base_counts = self.counts / effective_lengths
```
Comment (clintval) [inline edgePy/DGEList.py]: Have you confirmed this line of code does not double memory footprint?

File: edgePy/DGEList.py [hunk 13]
```diff
@@ -292,17 +292,40 @@ def rpkm(
...
+
+        self.counts = 10 ** 6 * base_counts / np.sum(base_counts, axis=0)[None, :]
```
Comment (clintval) [inline edgePy/DGEList.py]: Order of operations is clear but just to be safe I would group with `()`.


```python
self.counts = (10 ** 6) * base_counts / np.sum(base_counts, axis=0)[None, :]
```

Or better yet, drop the slicing and use `e`:

```python
self.counts = 10e6 * base_counts / np.sum(base_counts, axis=0)
```

Comment (clintval) [inline edgePy/DGEList.py]: However you are duplicating code from `DGEList.cpm()`. Please just call that method here instead of copy-paste the four lines. This reduces chances for divergent behavior and simplifies unit testing because if `cpm()` is well unit tested than you can focus more on the unit testing of `tpm`.

File: tests/test_DGEList.py [hunk 14]
```diff
@@ -203,11 +203,24 @@ def test_cpm():
     assert dge_list.counts[0][0] == first_pos * 1e6 / col_sum[0]
 
 
+def test_tpm():
+    dge_list = DGEList(filename=str(get_dataset_path(TEST_DATASET_NPZ)))
+    first_pos = dge_list.counts[0][0]
+    gene_lengths = np.random.randint(10, 1000, size=dge_list.genes.shape)
```
Comment (clintval) [inline tests/test_DGEList.py]: If you are using `random` please set the seed first so we can ensure reproducible test results.

File: edgePy/DGEList.py [hunk 15]
```diff
@@ -292,17 +292,40 @@ def rpkm(
...
+        if log:
+            self.counts[self.counts == 0] = prior_count
+            self.counts = np.log(self.counts)
```
Comment (clintval) [inline edgePy/DGEList.py]: IMHO you should just return self here to allow for chaining of functions in the future. Same for other norm. methods.

Like this theoretical:

```python
DGEList().from_something().rpkm().save_to_disk()
```

File: tests/test_DGEList.py [hunk 16]
```diff
@@ -203,11 +203,24 @@ def test_cpm():
...
+    desired = (
+        1e6
+        * (first_pos / gene_lengths[0])
+        / np.sum(dge_list.counts / gene_lengths[:, None], axis=0)[0]
+    )
```
Comment (clintval) [inline tests/test_DGEList.py]: Not bad, you are dynamically creating an expected result by copying some code from the main application.

This makes it hard to track down bugs with the logic since the logic exists in two places.

Norm. methods should probably be tested with a very small, hand-designed test case so you can explicitly write out all inputs, and all outputs.

File: tests/test_DGEList.py [hunk 17]
```diff
@@ -203,11 +203,24 @@ def test_cpm():
...
+    assert isinstance(first_pos, np.integer)
```
Comment (clintval) [inline tests/test_DGEList.py]: If `numpy` is imported I think `int` is implicitly cast to `np.integer`? So no need for the NumPy type. Can just use `int`.

File: edgePy/DGEList.py (hunk 2 above)
Comment (moritzschaefer) [reply edgePy/DGEList.py]: One is an array of shape (n,) the other one of shape (m,), the result needs to be a matrix of shape (n,m) as effective lengths depends on both sample and gene. 

I'll change for np.newaxis as it is more explicit and people directly know what to look/google for if they are not comfortable with numpy (https://stackoverflow.com/questions/944863/numpy-should-i-use-newaxis-or-none)

File: edgePy/DGEList.py (hunk 3 above)
Comment (moritzschaefer) [reply edgePy/DGEList.py]: Wanted to do that so I read the paper, but I was not 100% sure about their formulae as it includes one additional factor which the implementations apparently don't use
From publication:
image
Are you familiar with it and could elucidate it?

File: edgePy/DGEList.py (hunk 6 above)
Comment (moritzschaefer) [reply edgePy/DGEList.py]: Hmm, I only partially agree. What do we have typings for (i.e. people have the documentation inside their IDEs stating, that it's a boolean. If somebody on purpose wants to fuck it up, he for sure can).

File: edgePy/DGEList.py (hunk 15 above)
Comment (moritzschaefer) [reply edgePy/DGEList.py]: I think this is counter intuitive (as I suggested in the PR):
Either we have an imperative function rpkm (read "rpkm-normalize yourself!", which means that the object should transform/normalise itself) or we have a pure rpkm (read "return rpkm normalized counts" which means that the object doesn't modify itself). Or we have a Javascript like chaining function as you suggest.

Intuitively to me, the name rpkm suggests the second, I don't see a reason to adopt the method chaining pattern from Javascript here as it does not seem intuitively to me.

File: tests/test_DGEList.py (hunk 16 above)
Comment (moritzschaefer) [reply tests/test_DGEList.py]: You are definitely right, that this test can be improved. I adopted it from cpm_test tbh. 
I'll improve by writing and calculating a "hand calculated simple test"!

File: tests/test_DGEList.py (hunk 17 above)
Comment (moritzschaefer) [reply tests/test_DGEList.py]: Also adopted from cpm. I'll change in both

File: edgePy/DGEList.py (hunk 15 above)
Comment (clintval) [reply edgePy/DGEList.py]: That's OK. Mutating `self` and then returning `self` lends itself to chaining. 

IMHO if you are mutating self and returning `None` then returning `self` is free and let's the user of the API chain if they want. Not a dealbreaker for me but let's for pretty powerful queries like `object.filter().drop().normalize()`.

File: edgePy/DGEList.py (hunk 6 above)
Comment (clintval) [reply edgePy/DGEList.py]: Comes down to argument length vs readability so its up to the designer. I much prefer prepositions and verbs used to make a more readable expression at the sacrifice of a few more characters in typing. Examples for me are:


Conditional function: `maybe_do()`
Test membership: `has_object()`
Make a choice: `to_decide()`

With regard to typing. It is still a young feature in Python so used by developers using the library in their own. Many users of edgePy are likely to be working in Jupyter and type annotations are not going to help them.

If you keep argument name as a pure noun (`log`) it would be nice to have an early assertion validating it's type:

```python
if not isinstance(log, bool):
    raise TypeError('`log` must be of type bool.')
```

File: edgePy/DGEList.py (hunk 2 above)
Comment (clintval) [reply edgePy/DGEList.py]: Thanks! I had no idea one array was a row, and the other a column. A bit o' docs (in the args description) and an assertion of `np.array.shape()` will help.

I understand the need for slicing now! Would it be as cheap to transpose the second array?

```python
effective_lengths = (gene_lengths - mean_fragment_lengths.T).clip(min=1)
```

File: edgePy/DGEList.py (hunk 3 above)
Comment (clintval) [reply edgePy/DGEList.py]: Not terribly familiar, sorry.

I can only really advise on topics of API design, overall engineering, and scientific Python in general.

Maybe ping the Slack channel? I would honestly start with examining the implementation as exactly written in the source of `edgeR` to see how they compute.

Comment (clintval) [reply edgePy/DGEList.py]: Also! Formulae in docstrings corresponding to the algorithm as implemented would be amazing. If you know a bit of JS and want to add MathJax support for our Sphinx compiled docs I would be 👍 👍 👍 

http://www.sphinx-doc.org/en/master/usage/extensions/math.html

Comment (clintval) [reply edgePy/DGEList.py]: lmk if you need help with Sphinx, I use it (too frequently).

File: edgePy/DGEList.py (hunk 4 above)
Comment (clintval) [reply edgePy/DGEList.py]: See earlier comment. I missed that one was a 'row' and the other a 'column'.

File: tests/test_DGEList.py (hunk 14 above)
Comment (clintval) [reply tests/test_DGEList.py]: Along these lines check out plugins for harnessing the power of random:

https://pypi.org/project/pytest-randomly/

Or better yet, bounded random:

https://github.com/HypothesisWorks/hypothesis/tree/master/hypothesis-python

File: edgePy/DGEList.py (hunk 15 above)
Comment (clintval) [reply edgePy/DGEList.py]: However, I agree with the argument that a mutation of an object makes it so previous states are lost (at a savings of memory) and that may be an annoying behavior for scientific exploration.

Since these object can be quite large (thousands of genes and/or samples) I think mutating the original is fair but if a user wants to keep a previous state around, an explicit cast to a `.copy()` method should be made and the expectation would be a memory copy (deep copy).

I guess this would be a call for:

```python
class DGEList(object):
    
    def copy(self) -> 'DGEList':
        ...
        DGEList(self.counts.copy(), ...)
```

Comment (clintval) [reply edgePy/DGEList.py]: @apfejes what do you think?

File: edgePy/DGEList.py (hunk 2 above)
Comment (moritzschaefer) [reply edgePy/DGEList.py]: I think that would not work. You subtract an array of dimension (n,) minus an array dimension (m, 1). You need the np.newaxis (or None)

File: edgePy/DGEList.py (hunk 3 above)
Comment (moritzschaefer) [reply edgePy/DGEList.py]: I'll ask in Slack and check the edgeR source to verify. I'll have a look into the sphinx docs tomorrow or so.

File: edgePy/DGEList.py (hunk 6 above)
Comment (moritzschaefer) [reply edgePy/DGEList.py]: I agree that naming the variables appropriately is the best way. I'll change the name in all three functions to 'transform_to_log' and see if anybody complains. 
Another possible name would be apply_log_transform(ation)

File: edgePy/DGEList.py (hunk 7 above)
Comment (moritzschaefer) [reply edgePy/DGEList.py]: transform_to_log makes it clearer that it's a boolean IMO. Do you think it's appropriate?

File: edgePy/DGEList.py (hunk 13 above)
Comment (moritzschaefer) [reply edgePy/DGEList.py]: I agree for the three log_transform lines and will build a private log-function that will be applied by all three normalization methods. Before that I still need to check in the forum that normalization is applied correctly.

however self.counts = ... is different enough (i.e. I would not create an underlying private function that demonstrates that cpm and tpm have one code line relatively in common). 
One

File: edgePy/DGEList.py (hunk 15 above)
Comment (moritzschaefer) [reply edgePy/DGEList.py]: IMO either we should employ chaining as a paradigm for the whole library and implement it for every function (if possible) or not at all.
The copy function is necessary anyways. I think such a function is related to read/write_npz_file (as those functions access/store all data that would need to be copied.

File: tests/test_DGEList.py (hunk 16 above)
Comment (moritzschaefer) [reply tests/test_DGEList.py]: Done using hand written test and property validation

**Comment by apfejes** [comment]
Need to add:

`before_install:
  - sudo rm -f /etc/boto.cfg
`
to your .travis.yml file (right before install section).

in order for the unit tests to pass.

File: edgePy/DGEList.py (hunk 10 above)
Comment (apfejes) [reply edgePy/DGEList.py]: We should merge this branch with the af_rpkm branch to take advantage of the gene_lengths objets available there.

File: edgePy/DGEList.py (hunk 12 above)
Comment (apfejes) [reply edgePy/DGEList.py]: this can be easily cleaned up when we resolve #61

File: tests/test_DGEList.py [hunk 18]
```diff
@@ -72,16 +72,16 @@ def test_library_size():
...
+                63_579_607,
+                58_531_933,
+                39_138_521,
+                78_565_885,
+                48_667_119,
+                62_799_917,
+                66_032_107,
+                66_194_776,
+                55_085_875,
+                37_760_315,
```
Comment (apfejes) [inline tests/test_DGEList.py]: I don't get this change - these should be integers.
